library(Seurat)
library(tibble)
library(argparse)
# Basic Absolute and Global Scaling Vignette for Kim et al 2022
# Here we go through absolute and global scaling for the scRNA-seq "Mixology" ground truth dataset by Tian et al
# This dataset contains both ERCCs and UMIs


# Import matrices
# Matrics were generated by processing raw data from GSE118767 through a standard scPipe pipeline
# More details on scPipe found here: https://github.com/LuyiTian/scPipe

parser <- ArgumentParser()
parser$add_argument('filtered_feature_bc_matrix')
parser$add_argument('sample_id')
args <- parser$parse_args()

umi_raw <- Read10X(args$filtered_feature_bc_matrix)

# Create a Seurat object
UMI_mix <- CreateSeuratObject(counts = umi_raw, project = args$sample_id)
UMI_mix <- subset(UMI_mix, subset = nCount_RNA > 200) # filter out very low transcript cells

print(UMI_mix @assays$RNA)

# UMI Absolute Scaling
# To perform absolute scaling, we use the raw reads by only applying the log scaling step
print(sum(LayerData(UMI_mix, assay = "RNA", layer = "counts")[, "AAACCTGAGCTAGCCC-1"]))
UMI_mix@assays$RNA@layers$data <- as.matrix(log2(LayerData(UMI_mix, assay = "RNA", layer = "counts") + 1))
print(head(UMI_mix@assays$RNA@layers$data))
UMI_mix@meta.data$scaledcounts <- colSums(UMI_mix@assays$RNA@layers$data)

UMI_mix@meta.data <- rownames_to_column(UMI_mix@meta.data, var = "cell_barcode")


write.table(UMI_mix@meta.data,
            file = paste0(args$sample_id, "_absolute_counts.tsv"),
            sep = "\t",
            quote = FALSE,
            row.names = FALSE,
            col.names = TRUE)
